src/bootstrap/lib.rs

```diff
@@ -1841,3 +1849,13 @@ fn envify(s: &str) -> String {
         .flat_map(|c| c.to_uppercase())
         .collect()
 }
+
+pub fn find_recent_config_change_ids(current_id: usize) -> Vec<usize> {
+    // If someone accidentally enters an invalid value, we are checking its validity to
+    // ensure they do not miss major/breaking change logs.
```

Comment:
The value here seems pretty low for the cost of the list - it seems pretty unlikely that a typo will land on a PR that is the "latest" major change, so the user will keep seeing the warning continuously pretty much, right?

I'm fine with keeping it for now but we may want to reconsider if that proves a problem.

Reply:
Problem is, if the value is too high(e.g., writing 125898 instead of 115898 with a one char typo), there will be no logs until we reach that number on the rust-lang/rust PRs(because we only log for the IDs that are higher than the ID from the config.toml), which may take a year or so.

This is a problem which we should consider IMO.

Reply:
We log for inequality, no?

> &id != latest_config_id

So I'm not sure I follow that problem -- and if we didn't log for inequality, that would seem like it would work too.

Reply:
When `&id != latest_config_id` is true, then we collect IDs(calling `find_recent_config_change_ids(id)`) to log PRs that caused breaking changes. In find_recent_config_change_ids, if the ID is invalid, we call the panic right away. Because we know if value is invalid, it may be too high which will cause the problem I explained above, or it's too low, we will log all the PRs with saying "check these to see breaking changes, but the list will be incorrect. So, for invalid ID we may log all the PRs, or we may log none of the PR, or we may miss some PRs to be logged, etc.

Reply:
But PR IDs aren't assigned at merge time, so any sequential ordering is purely coincidental, right? In practice I expect we could use a random ID here and it would work equally well - the average user probably either ignores this warning or just bumps past it, looking at the git changelog if needed.

I think if the value isn't the latest value we basically can just say "look at the git commit history".

Reply:
> But PR IDs aren't assigned at merge time, so any sequential ordering is purely coincidental, right?

Not really. When a PR is opened, if config.rs or config.example.toml was edited rustbot will ping the author to add the PR ID to the related place(very similar to updating the 'download-ci-llvm-stamp'). Because PR IDs are generated incrementally, checking if there are higher IDs than the current ID should work. If we don't want to trust Github's ID generation right away. I can change it to split the ID array from the current ID position and log the IDs that come after the current ID(This should work too because we will always add PR IDs to the end of the ID list).

> In practice I expect we could use a random ID here and it would work equally well

This is quite different from what this PR provides; it is mostly intended for other teams. It will help them quickly review changes by providing the PRs that caused breaking changes. If you think this is unnecessary, I am fine with reverting this.

Reply:
> Because PR IDs are generated incrementally, checking if there are higher IDs than the current ID should work.

But they're generated at PR open time, not at merge time. So the CONFIG_CHANGE_HISTORY might consist of: `1, 100, 3, 10, 40, 2, 200` quite reasonably (in chronological order of merge times) -- so the filter by `>` would emit a subset, but essentially a random subset, of the list. We could try to find the ID that they *currently* have in the list and then log all subsequent IDs in the list, but that's not what the code is doing right now.

Reply:
Yeah you are right, I missed that part. https://github.com/rust-lang/rust/compare/3e878ef36c3dac97ce6107f1e819776568397119..7d3dcd918eb1f1308a1d901821ede3041d563ae9 should handle such cases now.